### ISSUE TITLE
RemoveLink appears not to work in assign.scm Scheme example

### ISSUE THREAD

**Issue by pennachin** [opencog/atomspace#281]
Running this example in guile:

```
scheme@(guile-user)> (use-modules (opencog exec))
scheme@(guile-user)> (define (show-eval-links)
...         (cog-map-type (lambda (h) (display h) #f) 'EvaluationLink))
scheme@(guile-user)> (define to-be-added
...         (InsertLink
...                 (TypeNode "EvaluationLink")
...                 (PredicateNode "some property")
...                 (ListLink
...                         (ConceptNode "thing A")
...                         (ConceptNode "B-dom-ness"))))
scheme@(guile-user)> (show-eval-links)
$1 = #f
scheme@(guile-user)> (cog-execute! to-be-added)
$2 = (EvaluationLink (av 0 0 0)
   (PredicateNode "some property")
   (ListLink
      (ConceptNode "thing A")
      (ConceptNode "B-dom-ness")
   )
)

scheme@(guile-user)> (show-eval-links)
(EvaluationLink (av 0 0 0)
   (PredicateNode "some property")
   (ListLink
      (ConceptNode "thing A")
      (ConceptNode "B-dom-ness")
   )
)
$3 = #f
scheme@(guile-user)> (define to-be-removed
...         (RemoveLink
...                 (TypeNode "EvaluationLink")
...                 (PredicateNode "some property")
...                 (ListLink
...                         (ConceptNode "thing A")
...                         (ConceptNode "B-dom-ness"))))
scheme@(guile-user)> (cog-execute! to-be-removed)
$4 = #<Undefined atom handle>

scheme@(guile-user)> (show-eval-links)
(EvaluationLink (av 0 0 0)
   (PredicateNode "some property")
   (ListLink
      (ConceptNode "thing A")
      (ConceptNode "B-dom-ness")
   )
)
$5 = #f
```

The comments in the source suggest that the final call to `(show-eval-links)` shouldn't include that `EvaluationLink`.

**Comment by pennachin** [comment]
Note: I'm running as of 00300be, a one line change to include module `(opencog exec)`. See #297

**Comment by linas** [comment]
It would seem that RemoveLink is broken, and that there is no unit test for it (there is one for InsertLink and AssignLink.)

Now that we have PutLink, I'm thinking that maybe all three of these are obsolete...

**Comment by linas** [comment]
well, I have a fix for this, but github is broken again, I can't merge it.  

github was working earlier today.

The breakage is the same as on Sunday, I can create pull requests, but am unable to merge; the merge button is grayed out.  Arghhh.  

https://status.github.com/   claims github is up.
https://status.github.com/messages  claims github is up.
https://twitter.com/githubstatus?lang=en  claims github is up.

This is a messed up bug.

**Comment by linas** [comment]
So when I do this, I get this messsage.
```
git push git@github.com:/opencog/atomspace remo:master
Total 0 (delta 0), reused 0 (delta 0)
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Required status check "ci/circleci" is pending
To git@github.com:/opencog/atomspace
 ! [remote rejected] remo -> master (protected branch hook declined)
error: failed to push some refs to 'git@github.com:/opencog/atomspace'
```

This is came error as Sunday.  I'm wasting time trying to figure out what's broken, I'm gonna try to disable CircleCI.  We can't use CircleCI if its broken all the time.

**Comment by pennachin** [comment]
Linas, what you are seeing is continuous integration working, not a bug. There is nothing broken. Every pull request triggers a CircleCI build. GitHub will not allow the pull request to be merged until CircleCI completes the build and it passes the tests. Simple as that, working as designed.

**Comment by linas** [comment]
Except it doesn't work!  

If you could see what I'm seeing, it would be clear.   Its not m,y rowser, cause I'm getting the same error from the command line!

**Comment by pennachin** [comment]
Linas, please listen to me. I can see what you see. It says a required status check is pending. 

That required status check is the CircleCI build. CircleCI is setup to prevent merging until the build is completed and tests pass. So a PR has to wait for that process, which takes 17-18 minutes for the atomspace repo.

Once the process is finished the PR can be merged. This is what happened to #282 (and its clones, the build is based on the commit), which I have just reopened.

Now that the build is done, you'll see that #282 states that the checks pass, and it can be merged.

**Comment by linas** [comment]
Yes, sure, I suppose that a circleci build might take 18 minutes. But there is no plausible reason for that build to slow down or stall a merge request.  Its not like we have to wait for that build to finish!  Something about the circleCI configuration is borken.  I'm looking at ti now, but its a bunch of checkboxes, and I'm too exhausted to read the docs.  I wasted a lot of time on this over the weekend.  It sucks.  I don't have time to screw around with this stuff.  It does not work well; its not a reliable tool.

**Comment by pennachin** [comment]
But that wait is the whole point of using CircleCI on merge requests: to get a passing build before the merge. Yes, it is like we have to wait for that build to finish, that is how continuous integration works. 

It works well and reliably: submit PR, build is triggered, merging is blocked until build passes, merging is unblocked after build passes. This is what we signed up for when we agreed to deploy continuous integration. 

If you still believe that this isn't working you're likely to keep wasting time. GitHub isn't broken, nor is CircleCI. If any part of this is still puzzling, let's discuss over email.

**Comment by linas** [comment]
I am completely puzzled why it should take 18 minutes.   It didn't take 18 minutes last week.  This week, it does. This is an unexplained, unexplicable change. Its not acceptable for a tool such as git to take 18 minutes to manage a merge request.   I'm using the word "its fucked up" because it is.   You can continously integrate till the cows come home, but there is no reason why a merge cannot be instantaneous.

**Comment by bgoertzel** [comment]
From what I understand, what you're seeing is an intentional consequence of how Amen and Cassio set up CircleCI.   They set it up to protect against merges that break the build, by automatically building each time a PR is made...

This will guarantee the branch always builds, right?   

I can see however that this way of using CircleCI does not work well for people who want to work by frequently merging small bits of code into the master branch.   It *would* work well for people who want to first get a somewhat larger chunk of functionality working in a feature branch, and then merge the feature branch into master, say, once a day or every few days or whatever....

So I guess this ties in indirectly with the "software process" discussion on the mailing list, right?   If it's considered desirable to work via frequently merging small bits of code into the master, then using CircleCI on the master in this way is not a good idea...

Using CircleCI in this way seems compatible with either

1) a single-mainline-branch workflow in which people do most of their work in feature branches and then periodically merge into master

or

2) a dual-mainline-branch workflow in which people can merge more freely into the develop branch (and there is no automatic build upon merge), but the "stable" branch is more protected, and e.g. has CircleCI configured to automatically check buildability after each merge

... ben

**Comment by pennachin** [comment]
Fixed by 63b584da7abfdc89095ddd89f05661f63fb8b4a8
